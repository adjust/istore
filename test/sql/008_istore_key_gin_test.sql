-- bigistore gin access should do index scan using gin;
CREATE TABLE t AS
SELECT i, string_agg(j::text || '=>0', ',')::bigistore AS s
FROM
generate_series(0, 10000) AS i,
generate_series(i, i + (i % 10)) AS j
GROUP BY i;
CREATE INDEX ON t USING gin(s);
SET enable_seqscan = 0;
EXPLAIN (COSTS OFF) SELECT * FROM t WHERE s ? 3;
-- bigistore gin access should find the matching rows for key 300;
SELECT s FROM t WHERE s ? 300 ORDER BY i;
-- bigistore gin access should find the matching rows for key 600;
SELECT s FROM t WHERE s ? 600 ORDER BY i;
-- bigistore gin access should find the matching rows for key 900;
SELECT s FROM t WHERE s ? 900 ORDER BY i;
-- istore gin access should do index scan using gin;
EXPLAIN (COSTS OFF) SELECT * FROM t WHERE s ? 3;
-- istore gin access should find the matching rows for key 300;
SELECT s FROM t WHERE s ? 300 ORDER BY i;
-- istore gin access should find the matching rows for key 600;
SELECT s FROM t WHERE s ? 600 ORDER BY i;
-- istore gin access should find the matching rows for key 900;
SELECT s FROM t WHERE s ? 900 ORDER BY i;
RESET enable_seqscan;
