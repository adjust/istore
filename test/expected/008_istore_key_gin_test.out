-- bigistore gin access should do index scan using gin;
CREATE TABLE t1 AS
SELECT i, string_agg(j::text || '=>0', ',')::bigistore AS s
FROM
generate_series(0, 10000) AS i,
generate_series(i, i + (i % 10)) AS j
GROUP BY i;
CREATE INDEX ON t1 USING gin(s);;
SET enable_seqscan = 0;;
EXPLAIN (COSTS OFF) SELECT * FROM t1 WHERE s ? 3;;
             QUERY PLAN              
-------------------------------------
 Bitmap Heap Scan on t1
   Recheck Cond: (s ? 3)
   ->  Bitmap Index Scan on t1_s_idx
         Index Cond: (s ? 3)
(4 rows)

RESET enable_seqscan;;
-- bigistore gin access should find the matching rows for key 300;
CREATE TABLE t2 AS
SELECT i, string_agg(j::text || '=>0', ',')::bigistore AS s
FROM
generate_series(0, 10000) AS i,
generate_series(i, i + (i % 10)) AS j
GROUP BY i;
CREATE INDEX ON t2 USING gin(s);;
SET enable_seqscan = 0;;
SELECT s FROM t2 WHERE s ? 300 ORDER BY i;;
                                                           s                                                            
------------------------------------------------------------------------------------------------------------------------
 "295"=>"0", "296"=>"0", "297"=>"0", "298"=>"0", "299"=>"0", "300"=>"0"
 "296"=>"0", "297"=>"0", "298"=>"0", "299"=>"0", "300"=>"0", "301"=>"0", "302"=>"0"
 "297"=>"0", "298"=>"0", "299"=>"0", "300"=>"0", "301"=>"0", "302"=>"0", "303"=>"0", "304"=>"0"
 "298"=>"0", "299"=>"0", "300"=>"0", "301"=>"0", "302"=>"0", "303"=>"0", "304"=>"0", "305"=>"0", "306"=>"0"
 "299"=>"0", "300"=>"0", "301"=>"0", "302"=>"0", "303"=>"0", "304"=>"0", "305"=>"0", "306"=>"0", "307"=>"0", "308"=>"0"
 "300"=>"0"
(6 rows)

RESET enable_seqscan;;
-- bigistore gin access should find the matching rows for key 600;
CREATE TABLE t3 AS
SELECT i, string_agg(j::text || '=>0', ',')::bigistore AS s
FROM
generate_series(0, 10000) AS i,
generate_series(i, i + (i % 10)) AS j
GROUP BY i;
CREATE INDEX ON t3 USING gin(s);;
SET enable_seqscan = 0;;
SELECT s FROM t3 WHERE s ? 600 ORDER BY i;;
                                                           s                                                            
------------------------------------------------------------------------------------------------------------------------
 "595"=>"0", "596"=>"0", "597"=>"0", "598"=>"0", "599"=>"0", "600"=>"0"
 "596"=>"0", "597"=>"0", "598"=>"0", "599"=>"0", "600"=>"0", "601"=>"0", "602"=>"0"
 "597"=>"0", "598"=>"0", "599"=>"0", "600"=>"0", "601"=>"0", "602"=>"0", "603"=>"0", "604"=>"0"
 "598"=>"0", "599"=>"0", "600"=>"0", "601"=>"0", "602"=>"0", "603"=>"0", "604"=>"0", "605"=>"0", "606"=>"0"
 "599"=>"0", "600"=>"0", "601"=>"0", "602"=>"0", "603"=>"0", "604"=>"0", "605"=>"0", "606"=>"0", "607"=>"0", "608"=>"0"
 "600"=>"0"
(6 rows)

RESET enable_seqscan;;
-- bigistore gin access should find the matching rows for key 900;
CREATE TABLE t4 AS
SELECT i, string_agg(j::text || '=>0', ',')::bigistore AS s
FROM
generate_series(0, 10000) AS i,
generate_series(i, i + (i % 10)) AS j
GROUP BY i;
CREATE INDEX ON t4 USING gin(s);;
SET enable_seqscan = 0;;
SELECT s FROM t4 WHERE s ? 900 ORDER BY i;;
                                                           s                                                            
------------------------------------------------------------------------------------------------------------------------
 "895"=>"0", "896"=>"0", "897"=>"0", "898"=>"0", "899"=>"0", "900"=>"0"
 "896"=>"0", "897"=>"0", "898"=>"0", "899"=>"0", "900"=>"0", "901"=>"0", "902"=>"0"
 "897"=>"0", "898"=>"0", "899"=>"0", "900"=>"0", "901"=>"0", "902"=>"0", "903"=>"0", "904"=>"0"
 "898"=>"0", "899"=>"0", "900"=>"0", "901"=>"0", "902"=>"0", "903"=>"0", "904"=>"0", "905"=>"0", "906"=>"0"
 "899"=>"0", "900"=>"0", "901"=>"0", "902"=>"0", "903"=>"0", "904"=>"0", "905"=>"0", "906"=>"0", "907"=>"0", "908"=>"0"
 "900"=>"0"
(6 rows)

RESET enable_seqscan;;
-- istore gin access should do index scan using gin;
CREATE TABLE t5 AS
SELECT i, string_agg(j::text || '=>0', ',')::istore AS s
FROM
generate_series(0, 10000) AS i,
generate_series(i, i + (i % 10)) AS j
GROUP BY i;
CREATE INDEX ON t5 USING gin(s);;
SET enable_seqscan = 0;;
EXPLAIN (COSTS OFF) SELECT * FROM t5 WHERE s ? 3;;
             QUERY PLAN              
-------------------------------------
 Bitmap Heap Scan on t5
   Recheck Cond: (s ? 3)
   ->  Bitmap Index Scan on t5_s_idx
         Index Cond: (s ? 3)
(4 rows)

RESET enable_seqscan;;
-- istore gin access should find the matching rows for key 300;
CREATE TABLE t6 AS
SELECT i, string_agg(j::text || '=>0', ',')::istore AS s
FROM
generate_series(0, 10000) AS i,
generate_series(i, i + (i % 10)) AS j
GROUP BY i;
CREATE INDEX ON t6 USING gin(s);;
SET enable_seqscan = 0;;
SELECT s FROM t6 WHERE s ? 300 ORDER BY i;;
                                                           s                                                            
------------------------------------------------------------------------------------------------------------------------
 "295"=>"0", "296"=>"0", "297"=>"0", "298"=>"0", "299"=>"0", "300"=>"0"
 "296"=>"0", "297"=>"0", "298"=>"0", "299"=>"0", "300"=>"0", "301"=>"0", "302"=>"0"
 "297"=>"0", "298"=>"0", "299"=>"0", "300"=>"0", "301"=>"0", "302"=>"0", "303"=>"0", "304"=>"0"
 "298"=>"0", "299"=>"0", "300"=>"0", "301"=>"0", "302"=>"0", "303"=>"0", "304"=>"0", "305"=>"0", "306"=>"0"
 "299"=>"0", "300"=>"0", "301"=>"0", "302"=>"0", "303"=>"0", "304"=>"0", "305"=>"0", "306"=>"0", "307"=>"0", "308"=>"0"
 "300"=>"0"
(6 rows)

RESET enable_seqscan;;
-- istore gin access should find the matching rows for key 600;
CREATE TABLE t7 AS
SELECT i, string_agg(j::text || '=>0', ',')::istore AS s
FROM
generate_series(0, 10000) AS i,
generate_series(i, i + (i % 10)) AS j
GROUP BY i;
CREATE INDEX ON t7 USING gin(s);;
SET enable_seqscan = 0;;
SELECT s FROM t7 WHERE s ? 600 ORDER BY i;;
                                                           s                                                            
------------------------------------------------------------------------------------------------------------------------
 "595"=>"0", "596"=>"0", "597"=>"0", "598"=>"0", "599"=>"0", "600"=>"0"
 "596"=>"0", "597"=>"0", "598"=>"0", "599"=>"0", "600"=>"0", "601"=>"0", "602"=>"0"
 "597"=>"0", "598"=>"0", "599"=>"0", "600"=>"0", "601"=>"0", "602"=>"0", "603"=>"0", "604"=>"0"
 "598"=>"0", "599"=>"0", "600"=>"0", "601"=>"0", "602"=>"0", "603"=>"0", "604"=>"0", "605"=>"0", "606"=>"0"
 "599"=>"0", "600"=>"0", "601"=>"0", "602"=>"0", "603"=>"0", "604"=>"0", "605"=>"0", "606"=>"0", "607"=>"0", "608"=>"0"
 "600"=>"0"
(6 rows)

RESET enable_seqscan;;
-- istore gin access should find the matching rows for key 900;
CREATE TABLE t8 AS
SELECT i, string_agg(j::text || '=>0', ',')::istore AS s
FROM
generate_series(0, 10000) AS i,
generate_series(i, i + (i % 10)) AS j
GROUP BY i;
CREATE INDEX ON t8 USING gin(s);;
SET enable_seqscan = 0;;
SELECT s FROM t8 WHERE s ? 900 ORDER BY i;;
                                                           s                                                            
------------------------------------------------------------------------------------------------------------------------
 "895"=>"0", "896"=>"0", "897"=>"0", "898"=>"0", "899"=>"0", "900"=>"0"
 "896"=>"0", "897"=>"0", "898"=>"0", "899"=>"0", "900"=>"0", "901"=>"0", "902"=>"0"
 "897"=>"0", "898"=>"0", "899"=>"0", "900"=>"0", "901"=>"0", "902"=>"0", "903"=>"0", "904"=>"0"
 "898"=>"0", "899"=>"0", "900"=>"0", "901"=>"0", "902"=>"0", "903"=>"0", "904"=>"0", "905"=>"0", "906"=>"0"
 "899"=>"0", "900"=>"0", "901"=>"0", "902"=>"0", "903"=>"0", "904"=>"0", "905"=>"0", "906"=>"0", "907"=>"0", "908"=>"0"
 "900"=>"0"
(6 rows)

RESET enable_seqscan;;
